src/OpenDiffix.Core/Analyzer.fs

```diff
@@ -304,35 +304,56 @@ let private rewriteToDiffixAggregate aidColumnsExpression query =
 
   query |> map exprMapper
 
+let private bucketExpand aidColumnsExpression =
+  let bucketCount = Expression.makeAggregate DiffixCount [ aidColumnsExpression ]
+  Expression.makeSetFunction GenerateSeries [ bucketCount ]
+
 let private addLowCountFilter aidColumnsExpression selectQuery =
   let lowCountFilter =
     Expression.makeAggregate DiffixLowCount [ aidColumnsExpression ]
     |> Expression.makeNot
 
-  if List.isEmpty selectQuery.GroupBy then
-    let selectedExpressions =
-      selectQuery.TargetList
-      |> List.map (fun selectedColumn -> selectedColumn.Expression)
-
-    if selectedExpressions |> List.forall Expression.isScalar then
-      // No need to group implicitly by constants, which are invalid bucket definitions anyway.
-      let implicitGroupExpressions = selectedExpressions |> List.filter (Expression.isConstant >> not)
-
-      // Non-grouping & non-aggregating query; group implicitly and expand
-      let bucketCount = Expression.makeAggregate DiffixCount [ aidColumnsExpression ]
-      let bucketExpand = Expression.makeSetFunction GenerateSeries [ bucketCount ]
-
-      { selectQuery with
-          TargetList =
-            { Expression = bucketExpand; Alias = ""; Tag = JunkTargetEntry }
-            :: selectQuery.TargetList
-          GroupBy = implicitGroupExpressions
-          Having = Expression.makeAnd lowCountFilter selectQuery.Having
-      }
-    else
-      // Non-grouping aggregate query; do nothing
-      selectQuery
-  else
+  let selectedExpressions =
+    selectQuery.TargetList
+    |> List.map (fun selectedColumn -> selectedColumn.Expression)
+
+  let nonConstantExpressions = selectedExpressions |> List.filter (Expression.isConstant >> not)
+
+  let isGrouping = List.isEmpty selectQuery.GroupBy |> not
+  let isAggregate = selectedExpressions |> List.forall Expression.isScalar |> not
+  let onlyConstantsSelected = List.isEmpty nonConstantExpressions
+
+  match (isGrouping, isAggregate, onlyConstantsSelected) with
```

Comment:
This is so much harder to parse now...

Reply:
harder than which version? Note, that this commit adds a 3rd layer of nested if. I started off with nested ifs and it was much worse off. So comparing against master isn't fair.

Also current refactored form comes from our convo with Edon, where we settled on this being a 4th top-level path which should be handled, so the `match with` version captures it - 4 separate top-level paths.

I needed to translate this into the `pg_diffix` implementation, I can't imagine doing this without the explicit, 4-legged match.

Reply:
The ifs don't all have to be nested, most could be on the same level. What do you think about this:
```F#
if not isGrouping && not isAggregate then // no aggregation
  let havingClause =
    if onlyConstantsSelected then
      Expression.makeAnd lowCountFilter selectQuery.Having
    else
      query.Having
  ...
else if not isGrouping && isAggregate then // global aggregation
   ...
else // grouped aggregation
  ...
```

Reply:
My main problem with the current version is that I have to scroll up at the beginning of the match when trying to understand each branch.

Reply:
This could work, I thought you were referring to the entire combined refactor. Let's do the `else if` version, however I still prefer the flatness of the `match`, but that's just personal preference.

I've also seen an approach with boolean aliases like `let GROUPING, NON_GROUPING = true, false`, I'm still not sure if I like it though.

Reply:
Nested if is done in 654eb5b7, let me know if you like it better than a `match with` with the boolean aliases for example

Reply:
Yes, I think it looks better now.